Adding support for Simba JDBC Driver

rhou1 opened the pull request:
Hi Abhishek,

Can you take a look at these changes?

I modified the json files to include a new category for simba jdbc tests.  I also added code to the test framework to handle json files that have not been updated to include the new category.  We can discuss which way is better, or have both options.

I will send the json file changes and test result changes in a separate pull request.

Thanks.

--Robert

Agirish commented:
Thanks Robert. I'm taking a look now. 

@cchang738, can you also please review?

Agirish commented on framework/src/main/java/org/apache/drill/test/framework/ColumnList.java:
```diff
@@ -88,8 +93,36 @@ public int hashCode() {
+        // order
+        if ((s1.length() > 0) &&
+           ((s1.charAt(0) == '{') || (s1.charAt(0) == '[')) ) {
+          s1 = s1.replace("\n","");
```
Have you considered replacing "\n" (new line) with " " (space) instead of "" (empty string)? I'm concerned some strings with new lines will end up concatenated instead of simple eliminating new lines. 
For ex: 
"How are
you" 
should not end up as "How areyou", but instead as "How are you".

rhou1 replied:
I don't think a JSON string can have a "\n" in it.  A JSON string can have "\\n" in it, but I believe JSON strings have to be "well-formed".

Agirish replied:
I'm still a little unclear on this. Have you considered JSON data with multi-line text? Are we replacing all newlines with ""?

rhou1 replied:
I don't think you can have a new line break embedded within a string.

http://stackoverflow.com/questions/2392766/multiline-strings-in-json

Yes, I'm replacing all newlines with "".  I could replace them with " ", but I'm not sure how much of a difference it would make.

rhou1 replied:
Perhaps you can give me an example of what you are thinking about, in case I mis-understand your question.

Agirish commented on framework/src/main/java/org/apache/drill/test/framework/ColumnList.java:
```diff
@@ -88,8 +93,36 @@ public int hashCode() {
       sb.append(values.get(i) + "\t");
     }
+    int type = (Integer) (types.get(values.size()-1));
+    if (categories.contains("smoke_simba_jdbc") && (type == Types.VARCHAR)) {
```
Same as above - replace text with constant.

rhou1 replied:
Sure.

Agirish commented on framework/src/main/java/org/apache/drill/test/framework/ConnectionPool.java:
```diff
@@ -19,6 +19,7 @@
 
 import com.google.common.collect.Queues;
 //import org.apache.drill.jdbc.Driver;
+//import com.mapr.drill.jdbc41.Driver;
```
Any reason for having commented out imports?

rhou1 replied:
The import is used for testing.  The line in question is for the Simba driver.  The line above it is for the Apache driver.

Agirish replied:
Okay thanks!

Agirish commented on framework/src/main/java/org/apache/drill/test/framework/DrillTestJdbc.java:
```diff
@@ -100,12 +100,15 @@ public void run() {
           matrix.verificationTypes.get(0).equalsIgnoreCase("filter-ratio")) {
         setTestStatus(testVerifier.verifyTextPlan(modeler.expectedFilename, outputFilename));
+      } else if (matrix.verificationTypes.get(0).equalsIgnoreCase("text") ) {
+        setTestStatus(testVerifier.verifyText(modeler.expectedFilename, outputFilename));
```
Is "text" a new verification type? Can we have more comments on each of the verification types?

rhou1 replied:
It is a new verification type.  I was originally going to use it for the JDBC tests, but I don't need it for that any more.  I am now planning to use it for orderby testing.

Agirish replied:
Can we separate this out into a new PR?

rhou1 replied:
Sure, I can pull the code out and put it in a new PR.  I just thought I would save some paperwork.

Agirish commented on framework/src/main/java/org/apache/drill/test/framework/Utils.java:
```diff
@@ -169,7 +170,21 @@ public static ConnectionPool getConnectionPool() {
+              //    smoke_simba_jdbc tests
+              // therefore assume this json file should be modified to include
+              //    smoke_simba_jdbc tests
+              // therefore check for smoke_simba_jdbc tests
```
Can we please clean-up the comments? And include clear details on each of the test definition file types (apache.json, simba.jdbc.json, etc.)? 

Also, may be consider renaming apache.json to either "apache.jdbc.json" or just the default ".json" to maintain backward compatibility.

Agirish replied:
Also can you please add comments on backward compatibility? Are we saying all existing json files must be renamed to end with apache.json?

rhou1 replied:
I thought it would be more helpful if the comments could explain the code step-by-step.  But I can write a general comment.

I was wondering where to document the test definition file types.  Perhaps this is a good place for it.

Yes, I can rename apache.json to apache.jdbc.json.

Backward compatibilty is maintained.  By default, our current json files stay.  I have added two more categories (for jdbc and odbc) to each json file.  Most tests pass without any changes for Apache and Simba drivers.  A small number do not.  For these tests, I have renamed the query files to use a new file extension (*.q2) and created new json files for them.  There is a json file that Apache uses for these special tests, and a json file that Simba uses for these special tests.  Apache expected results files remain as they were.  I added new Simba expected results files with a new file extension (*.sjdbc.e_tsv as an example).

Having said that, I am concerned it is error-prone to require new json files to have additional categories for Simba.  So I added this code to enable existing json files to run with Simba without any changes.  When we had our last discussions, this was considered not the best approach.  So I have implemented both, so we can see what the impact is.

Perhaps I should add a few json files and test files so you can see how the whole thing works together.  Initially I just put the code here because it seems useful to have code separated from the rest of the changes, when the changes are large.

Agirish commented on framework/src/main/java/org/apache/drill/test/framework/Utils.java:
```diff
@@ -169,7 +170,21 @@ public static ConnectionPool getConnectionPool() {
         for (String testGroup : testGroups) {
           if (categories != null && !categories.contains(testGroup)) {
-            continue;
+            if (categories.contains("smoke") &&
```
Let's not use "smoke" as the category, as it becomes restrictive and also "smoke" isn't well defined by us.

rhou1 replied:
Perhaps we should address this too, then.  I went through almost all of the json files, and I saw mostly smoke.  I saw a few negative and other ones, but very few.  I would be happy to go through them all again and make them more consistent if possible.

rhou1 replied:
Counted 321 test definition files.

droptable suite uses droptable (2 json files)
maprdr suite uses functional_maprdb (1 json file)
some aggregate tests use "functional,smoke" (4 json files, including 2 under limit0)
impersonation suite uses impersonation (15 json files)
some suites use negative (10 json files)
complex uses regression (9 json files)
the rest use smoke (280 json files)

rhou1 replied:
The reason "smoke" is used as a category here is to show another way to incorporate the Simba drivers.  The one that we agreed upon is to add the Simba categories to the json files.  But this requires constant maintenance as we add more json files.  At the time, it was preferred not to make more changes to the framework than necessary.

So I decided to prototype what change would be needed to automatically invoke the Simba categories without changing the json files, to better evaluate the complexity.  In this case, I'm using "smoke" as a reference.  If a json file uses the "smoke" category, the framework will automatically add the Simba category.

We can go either way.  I have made all the changes to the json files, which will be in a separate PR.

Agirish replied:
I think having a single way of defining "simba_jdbc" category is sufficient. Let's not use "smoke". Any test suite which needs to be added or updated for Simba JDBC should define the "simba_jdbc" category. I'm using it similarly for MapR-DB tests - they are categorized as "functional_maprdb".

rhou1 replied:
I think there is a difference between functional_maprdb and simba_jdbc.  Most tests should be run with simba_jdbc.  Most tests should not run with functional_maprdb.  I think the majority of the test suites should run with simba_jdbc.  So why do this manually?  Why not have the test framework handle this for us?

rhou1 replied:
We can have a group discussion about this when you get back.

rhou1 replied:
The main concern I have is, how do we know if we missed simba_jdbc somewhere.  I think it would be better to run with it "everywhere" and disable it when necessary.  Rather than add it when we remember to add it, which means we may miss it in some places.

rhou1 commented on framework/src/main/java/org/apache/drill/test/framework/ColumnList.java:
```diff
@@ -88,8 +93,36 @@ public int hashCode() {
     StringBuilder sb = new StringBuilder();
     for (int i = 0; i < values.size() - 1; i++) {
+      int type = (Integer) (types.get(i));
+      if (categories.contains("smoke_simba_jdbc") && (type == Types.VARCHAR)) {
```
Sure.  I was thinking that the Simba tests would be parallel to the smoke tests.

rhou1 commented:
Updated some code to address comments.

Added changes to table_function test suite as an example of how test definition files, query files and expected output files are changed.

rhou1 commented:
Removed code for new verifyText method.  Will create a new PR so it can be used with orderby tests.

cchang738 commented:
@rhou1 Let's discuss this PR in person. Thanks.

rhou1 commented:
Had discussions with Chun.  Re-open PR for further discussions with the group.

rhou1 commented:
Re-open pull request.

rhou1 commented:
Force-push latest content to bring pull request back to the current state.

cchang738 commented:
@rhou1 the proposed approach is rather messy. You need to move certain existing query files from .q to .q2; add a smoke-jdbc tag to certain .json files but not for some others; add another set of .json files to point to different baselines. Running tests with different drivers and maintaining tests can be confusing too.

Here is what I think we should do:

1. make no changes to query file names and test definition file (.json file)
2. add second sets of baseline with a predefined extension, such as *.e-mapr 
3. add a run time parameter such as -o to pick up an alternative baseline file if exists based on extension (such as *.e-mapr and can be specified at commandline).

This is hopefully a much simpler approach. When we need to test an alternative driver, just give the -o option and the framework will automatically pick the right baseline.

rhou1 commented:
@cchang738 
I used the approach in this PR to maintain the flexibility provided by the framework.  And this PR provides a way to avoid adding the smoke_simba_jdbc tag to certain json files (by running smoke_simba_jdbc tests wherever smoke tests are run).  Adding the smoke_simba_jdbc tag was the original design that we agreed to, but I agree it can be messy.  Which is why I added a way to avoid it.

Having said that, I think your approach can work too, although it is not as flexible in some ways.   And if we need the extra flexibility, we could use the approach in this PR for those specific cases.  So I'm happy to go with either approach.

rhou1 commented:
We also need to find a way to accommodate tests that cannot be run with Simba.  They need to be run with Apache, and disabled to run with Simba.  We might be able to use another file extension to identify these tests.

rhou1 commented:
Cleaned up the code based on group discussion some time ago.  Abhishek, Chun, please take a look.

rhou1 commented:
Added comments to clarify why queryFileExtension.substring(2).  Before this change, the queryFileExtention was "*.fail".  Now the queryFileExtension is "*.q.fail".  This enables the ".fail" extension to be used with other files, such as expected results files.  When an expected results file has a ".fail" extension, such as "*.e_tsv.sjdbc.fail", the corresponding test will not be executed with the Simba JDBC driver due to an existing bug.

rhou1 commented:
Added removeNewLines function to remove new line characters (\n) except when they exist in a string demarcated by double quotes.

This allows us to handle a case like this:
0: jdbc:drill:drillbit=10.10.100.186> select * from `newline5.csv`;
+--------------------------------+
|            columns             |
+--------------------------------+
| ["line1"]                      |
| ["{line2\nabc\ndef\n}"]        |
| ["line3"]                      |
| ["{line4\n   abc\n   def\n}"]  |
+--------------------------------+

This is a parquet file with strings.  Here, the string does not have double quotes.  The code cannot handle this case, but this is an issue for the test framework in general, because we need each record to be on a single line.
0: jdbc:drill:drillbit=10.10.100.186> select * from `newline7`;
+-------------------------+
|         EXPR$0          |
+-------------------------+
| line1                   |
| {line2
abc
def
}        |
| line3                   |
| {line4
   abc
   def
}  |
+-------------------------+

cchang738 commented:
+1 LGTM. But please run pre-commit test with latest change if you haven't done so.

Agirish commented:
+1. Let's get this in, as it's been pending from long. We can always work on improving anything on a later date.

rhou1 commented:
I re-ran the pre-commit tests with the latest changes.

rhou1 commented:
Chun,  Abhishek, thanks for the review.
